internal/command/updateimagetag.go

```diff
@@ -0,0 +1,190 @@
+	Run: func(ctx context.Context) error {
+		if !supportedLanguages[flagLanguage] {
+			return fmt.Errorf("invalid -language flag specified: %q", flagLanguage)
+		}
+		if flagPush && flagGitHubToken == "" {
+			return errors.New("-github-token must be provided if -push is set to true")
+		}
+		if flagTag == "" {
+			return errors.New("-tag is not provided")
+		}
+
+		startOfRun := time.Now()
+
+		// tmpRoot is a newly-created working directory under /tmp
+		// We do any cloning or copying under there.
+		tmpRoot, err := createTmpWorkingRoot(startOfRun)
+		if err != nil {
+			return err
+		}
+
+		var apiRepo *gitrepo.Repo
+		if flagAPIRoot == "" {
+			apiRepo, err = cloneGoogleapis(ctx, tmpRoot)
+			if err != nil {
+				return err
+			}
+		} else {
+			apiRoot, err := filepath.Abs(flagAPIRoot)
+			slog.Info(fmt.Sprintf("Using apiRoot: %s", apiRoot))
+			if err != nil {
+				slog.Info(fmt.Sprintf("Error retrieving apiRoot: %s", err))
+				return err
+			}
+			apiRepo, err = gitrepo.Open(ctx, apiRoot)
+			if err != nil {
+				return err
+			}
+			clean, err := gitrepo.IsClean(ctx, apiRepo)
+			if err != nil {
+				return err
+			}
+			if !clean {
+				return errors.New("api repo must be clean before updating the language image tag")
+			}
+		}
+
+		var outputDir string
+		if flagOutput == "" {
+			outputDir = filepath.Join(tmpRoot, "output")
+			if err := os.Mkdir(outputDir, 0755); err != nil {
+				return err
+			}
+			slog.Info(fmt.Sprintf("No output directory specified. Defaulting to %s", outputDir))
+		} else {
+			outputDir, err = filepath.Abs(flagOutput)
+			if err != nil {
+				return err
+			}
+		}
+
+		var languageRepo *gitrepo.Repo
+		if flagRepoRoot == "" {
+			languageRepo, err = cloneLanguageRepo(ctx, flagLanguage, tmpRoot)
+			if err != nil {
+				return err
+			}
+		} else {
+			repoRoot, err := filepath.Abs(flagRepoRoot)
+			if err != nil {
+				return err
+			}
+			languageRepo, err = gitrepo.Open(ctx, repoRoot)
+			if err != nil {
+				return err
+			}
+			clean, err := gitrepo.IsClean(ctx, apiRepo)
+			if err != nil {
+				return err
+			}
+			if !clean {
+				return errors.New("language repo must be clean before update")
+			}
+		}
+
+		state, err := loadState(languageRepo)
+		if err != nil {
+			return err
+		}
+
+		if state.ImageTag == flagTag {
+			return errors.New("specified tag is already in language repo state")
+		}
+		state.ImageTag = flagTag
+		image := deriveImage(state)
+		saveState(languageRepo, state)
+
+		// Take a defensive copy of the generator input directory from the language repo.
+		generatorInput := filepath.Join(tmpRoot, "generator-input")
+		if err := os.CopyFS(generatorInput, os.DirFS(filepath.Join(languageRepo.Dir, "generator-input"))); err != nil {
+			return err
+		}
+
+		// Perform "generate, clean" on each element in ApiGenerationStates.
+		for _, apiState := range state.ApiGenerationStates {
+			err = regenerateApi(ctx, apiRepo, languageRepo, generatorInput, image, outputDir, apiState)
```

Comment:
I wonder why to regenerate for each api instead of using container.generate() for all? Is it because we do not record a global equivalent to apiState.LastGeneratedCommit in the pipeline-state file?

Reply:
It's because we want to regenerate at whatever commit it was previously generated. It could be that "latest" is broken for some APIs, for example.